# Enefit - Predict Energy Behavior of Prosumers | Data Exploration | Merge

<img src="./header.png" width="1200" height="400"/>

# 1. Import the necessary libraries

In [1]:
import os
import pandas as pd
import numpy as np
import warnings
from tqdm.notebook import tqdm_notebook

warnings.filterwarnings('ignore')

# 2. Read the data

In [2]:
INPUT_DIR = '../../../Kaggle_input/input/Enefit-Predict-Energy-Behavior/predict-energy-behavior-of-prosumers/'

In [3]:
os.listdir(INPUT_DIR)

['county_id_to_name_map.json',
 'client.csv',
 'example_test_files',
 'enefit',
 'weather_station_to_county_mapping.csv',
 'train.csv',
 'public_timeseries_testing_util.py',
 'gas_prices.csv',
 'historical_weather.csv',
 'forecast_weather.csv',
 'electricity_prices.csv']

In [4]:
train = pd.read_csv(INPUT_DIR + 'train.csv', parse_dates=["datetime"], engine="python")
client = pd.read_csv(INPUT_DIR + 'client.csv', parse_dates=["date"], engine="python")
weather_map = pd.read_csv(INPUT_DIR + 'weather_station_to_county_mapping.csv', engine="python")
gas_price = pd.read_csv(INPUT_DIR + 'gas_prices.csv', parse_dates=["forecast_date", "origin_date"], engine="python")
historic_weather = pd.read_csv(INPUT_DIR + 'historical_weather.csv', parse_dates=["datetime"], engine="python")
forecast_weather = pd.read_csv(INPUT_DIR + 'forecast_weather.csv', parse_dates=["origin_datetime", "forecast_datetime"], engine="python")
electricity_price = pd.read_csv(INPUT_DIR + 'electricity_prices.csv', parse_dates=["forecast_date", "origin_date"], engine="python")

In [5]:
train.head()

county  is_business  product_type  target  is_consumption   datetime  \
0       0            0             1   0.713               0 2021-09-01   
1       0            0             1  96.590               1 2021-09-01   
2       0            0             2   0.000               0 2021-09-01   
3       0            0             2  17.314               1 2021-09-01   
4       0            0             3   2.904               0 2021-09-01   

   data_block_id  row_id  prediction_unit_id  
0              0       0                   0  
1              0       1                   0  
2              0       2                   1  
3              0       3                   1  
4              0       4                   2

In [6]:
client.head()

product_type  county  eic_count  installed_capacity  is_business  \
0             1       0        108              952.89            0   
1             2       0         17              166.40            0   
2             3       0        688             7207.88            0   
3             0       0          5              400.00            1   
4             1       0         43             1411.00            1   

        date  data_block_id  
0 2021-09-01              2  
1 2021-09-01              2  
2 2021-09-01              2  
3 2021-09-01              2  
4 2021-09-01              2

In [7]:
weather_map.sample(5)

county_name  longitude  latitude  county
70       Harjumaa       25.7      59.4     0.0
99            NaN       27.7      58.5     NaN
81        Võrumaa       26.7      57.9    15.0
76      Jõgevamaa       26.2      58.8     4.0
77  Lääne-Virumaa       26.2      59.1     5.0

In [8]:
gas_price.head()

forecast_date  lowest_price_per_mwh  highest_price_per_mwh origin_date  \
0    2021-09-01                 45.23                  46.32  2021-08-31   
1    2021-09-02                 45.62                  46.29  2021-09-01   
2    2021-09-03                 45.85                  46.40  2021-09-02   
3    2021-09-04                 46.30                  46.80  2021-09-03   
4    2021-09-05                 46.30                  46.58  2021-09-04   

   data_block_id  
0              1  
1              2  
2              3  
3              4  
4              5

In [9]:
historic_weather.head()

datetime  temperature  dewpoint  rain  snowfall  surface_pressure  \
0 2021-09-01         14.2      11.6   0.0       0.0            1015.9   
1 2021-09-01         13.9      11.5   0.0       0.0            1010.7   
2 2021-09-01         14.0      12.5   0.0       0.0            1015.0   
3 2021-09-01         14.6      11.5   0.0       0.0            1017.3   
4 2021-09-01         15.7      12.9   0.0       0.0            1014.0   

   cloudcover_total  cloudcover_low  cloudcover_mid  cloudcover_high  \
0                31              31               0               11   
1                33              37               0                0   
2                31              34               0                0   
3                 0               0               0                0   
4                22              25               0                0   

   windspeed_10m  winddirection_10m  shortwave_radiation  \
0       7.083333                  8                  0.0   
1       5.111111                359                  0.0   
2       6.333333                355                  0.0   
3       8.083333                297                358.0   
4       8.416667                  5                  0.0   

   direct_solar_radiation  diffuse_radiation  latitude  longitude  \
0                     0.0                0.0      57.6       21.7   
1                     0.0                0.0      57.6       22.2   
2                     0.0                0.0      57.6       22.7   
3                   277.0               81.0      57.6       23.2   
4                     0.0                0.0      57.6       23.7   

   data_block_id  
0            1.0  
1            1.0  
2            1.0  
3            1.0  
4            1.0

In [10]:
forecast_weather.head()

latitude  longitude     origin_datetime  hours_ahead  temperature  \
0      57.6       21.7 2021-09-01 02:00:00            1    15.655786   
1      57.6       22.2 2021-09-01 02:00:00            1    13.003931   
2      57.6       22.7 2021-09-01 02:00:00            1    14.206567   
3      57.6       23.2 2021-09-01 02:00:00            1    14.844507   
4      57.6       23.7 2021-09-01 02:00:00            1    15.293848   

    dewpoint  cloudcover_high  cloudcover_low  cloudcover_mid  \
0  11.553613         0.904816        0.019714        0.000000   
1  10.689844         0.886322        0.004456        0.000000   
2  11.671777         0.729034        0.005615        0.000000   
3  12.264917         0.336304        0.074341        0.000626   
4  12.458887         0.102875        0.088074        0.000015   

   cloudcover_total  10_metre_u_wind_component  10_metre_v_wind_component  \
0          0.905899                  -0.411328                  -9.106137   
1          0.886658                   0.206347                  -5.355405   
2          0.730499                   1.451587                  -7.417905   
3          0.385468                   1.090869                  -9.163999   
4          0.176590                   1.268481                  -8.975766   

   data_block_id   forecast_datetime  direct_solar_radiation  \
0              1 2021-09-01 03:00:00                     0.0   
1              1 2021-09-01 03:00:00                     0.0   
2              1 2021-09-01 03:00:00                     0.0   
3              1 2021-09-01 03:00:00                     0.0   
4              1 2021-09-01 03:00:00                     0.0   

   surface_solar_radiation_downwards  snowfall  total_precipitation  
0                                0.0       0.0                  0.0  
1                                0.0       0.0                  0.0  
2                                0.0       0.0                  0.0  
3                                0.0       0.0                  0.0  
4                                0.0       0.0                  0.0

In [11]:
electricity_price.head()

forecast_date  euros_per_mwh         origin_date  data_block_id
0 2021-09-01 00:00:00          92.51 2021-08-31 00:00:00              1
1 2021-09-01 01:00:00          88.90 2021-08-31 01:00:00              1
2 2021-09-01 02:00:00          87.35 2021-08-31 02:00:00              1
3 2021-09-01 03:00:00          86.88 2021-08-31 03:00:00              1
4 2021-09-01 04:00:00          88.43 2021-08-31 04:00:00              1

# 3. Explore individual datasets

## 3.1 Explore train data

In [12]:
train.tail()

county  is_business  product_type   target  is_consumption  \
2018347      15            1             0  197.233               1   
2018348      15            1             1    0.000               0   
2018349      15            1             1   28.404               1   
2018350      15            1             3    0.000               0   
2018351      15            1             3  196.240               1   

                   datetime  data_block_id   row_id  prediction_unit_id  
2018347 2023-05-31 23:00:00            637  2018347                  64  
2018348 2023-05-31 23:00:00            637  2018348                  59  
2018349 2023-05-31 23:00:00            637  2018349                  59  
2018350 2023-05-31 23:00:00            637  2018350                  60  
2018351 2023-05-31 23:00:00            637  2018351                  60

In [13]:
# Shape of data
train.shape

(2018352, 9)

* We have total **2018352** records and **9** columns in the data.

In [14]:
# Check all columns and there datatypes
train.dtypes

county                         int64
is_business                    int64
product_type                   int64
target                       float64
is_consumption                 int64
datetime              datetime64[ns]
data_block_id                  int64
row_id                         int64
prediction_unit_id             int64
dtype: object

In [15]:
# Missing values
train.isna().sum()

county                  0
is_business             0
product_type            0
target                528
is_consumption          0
datetime                0
data_block_id           0
row_id                  0
prediction_unit_id      0
dtype: int64

* We have **528** missing values in the target column, let's check these values

In [16]:
train[train.target.isna()]

county  is_business  product_type  target  is_consumption  \
178938        0            0             1     NaN               0   
178939        0            0             1     NaN               1   
178940        0            0             2     NaN               0   
178941        0            0             2     NaN               1   
178942        0            0             3     NaN               0   
...         ...          ...           ...     ...             ...   
1806379      15            1             0     NaN               1   
1806380      15            1             1     NaN               0   
1806381      15            1             1     NaN               1   
1806382      15            1             3     NaN               0   
1806383      15            1             3     NaN               1   

                   datetime  data_block_id   row_id  prediction_unit_id  
178938  2021-10-31 03:00:00             60   178938                   0  
178939  2021-10-31 03:00:00             60   178939                   0  
178940  2021-10-31 03:00:00             60   178940                   1  
178941  2021-10-31 03:00:00             60   178941                   1  
178942  2021-10-31 03:00:00             60   178942                   2  
...                     ...            ...      ...                 ...  
1806379 2023-03-26 03:00:00            571  1806379                  64  
1806380 2023-03-26 03:00:00            571  1806380                  59  
1806381 2023-03-26 03:00:00            571  1806381                  59  
1806382 2023-03-26 03:00:00            571  1806382                  60  
1806383 2023-03-26 03:00:00            571  1806383                  60  

[528 rows x 9 columns]

In [17]:
train[train.target.isna()]["datetime"].value_counts()

2022-10-30 03:00:00    136
2022-03-27 03:00:00    134
2023-03-26 03:00:00    132
2021-10-31 03:00:00    126
Name: datetime, dtype: int64

* We can see the target is missing for the hour **03:00:00**. We need to handle these values. Let's impute these values using ffill.

In [18]:
train.datetime.dtype

dtype('<M8[ns]')

In [19]:
train["date"] = pd.to_datetime(train["datetime"].dt.date)
train["time"] = train["datetime"].dt.time

In [20]:
train[["county", "is_business", "product_type", "is_consumption", "time", "date"]].duplicated().value_counts()

False    2018352
dtype: int64

In [21]:
missing_train = train.groupby(["county", "is_business", "product_type", "is_consumption", "time", "date"])["target"].mean().reset_index().ffill()

In [22]:
missing_train.tail()

county  is_business  product_type  is_consumption      time  \
2018347      15            1             3               1  23:00:00   
2018348      15            1             3               1  23:00:00   
2018349      15            1             3               1  23:00:00   
2018350      15            1             3               1  23:00:00   
2018351      15            1             3               1  23:00:00   

              date   target  
2018347 2023-05-27  183.756  
2018348 2023-05-28  161.650  
2018349 2023-05-29  177.056  
2018350 2023-05-30  200.718  
2018351 2023-05-31  196.240

In [23]:
train.drop(columns=["target"], inplace=True)
train = train.merge(
    missing_train,
    on=["county", "is_business", "product_type", "is_consumption", "time", "date"],
    suffixes=("", "_train")
)
train.shape

(2018352, 11)

In [24]:
train.isna().sum()

county                0
is_business           0
product_type          0
is_consumption        0
datetime              0
data_block_id         0
row_id                0
prediction_unit_id    0
date                  0
time                  0
target                0
dtype: int64

* No missing data anymore.

In [25]:
train[train.target.isna()]["datetime"].value_counts()

Series([], Name: datetime, dtype: int64)

In [26]:
# Check the date present
train_start_date = train.date.min()
train_end_date = train.date.max()
print(train_start_date, train_end_date)
# Create a date range
train_date_range = pd.date_range(start=train_start_date, end=train_end_date, freq='D')

2021-09-01 00:00:00 2023-05-31 00:00:00


* We have data from **'2021-09-01 00:00:00** to **2023-05-31 23:00:00**

In [27]:
# Check for missing dates
missing_dates = train_date_range[~train_date_range.isin(train.date)]

# Print the missing dates
print("Missing Dates:")
print(missing_dates)

Missing Dates:
DatetimeIndex([], dtype='datetime64[ns]', freq='D')


* Nice we have no missing dates. Let's check for missing time as well as we have hourly data.

In [28]:
# Check the datetime present
start_datetime = train.datetime.min()
end_datetime = train.datetime.max()
print(start_datetime, end_datetime)
# Create a date range
datetime_range = pd.date_range(start=start_datetime, end=end_datetime, freq='H')

2021-09-01 00:00:00 2023-05-31 23:00:00


In [29]:
# Check for missing dates
missing_datetimes = datetime_range[~datetime_range.isin(train.datetime)]

# Print the missing dates
print("Missing Dates:")
print(missing_datetimes)

Missing Dates:
DatetimeIndex([], dtype='datetime64[ns]', freq='H')


* Great!! No time is missing as well.

In [30]:
# Counties
train.county.value_counts().sort_index()

0     212928
1      91872
2     115200
3     122496
4     147264
5     151632
6      30624
7     173088
8      91872
9     122496
10    134640
11    198000
12     30624
13    121056
14    125808
15    148752
Name: county, dtype: int64

## 3.2 Explore the client data

In [31]:
# Shape
client.shape

(41919, 7)

* We have total **41919** reccords and **7** columns in the client data.

In [32]:
# Chec kall columns and there data types
client.dtypes

product_type                   int64
county                         int64
eic_count                      int64
installed_capacity           float64
is_business                    int64
date                  datetime64[ns]
data_block_id                  int64
dtype: object

In [33]:
# Missing values
client.isna().sum()

product_type          0
county                0
eic_count             0
installed_capacity    0
is_business           0
date                  0
data_block_id         0
dtype: int64

* No missing values in the client data.

In [34]:
# Define the start and end dates based on your data
start_date = client.date.min()
end_date = client.date.max()
print(start_date, end_date)
# Create a date range
date_range = pd.date_range(start=start_date, end=end_date, freq='D')

2021-09-01 00:00:00 2023-05-29 00:00:00


* We have client data till **2023-05-29**, we will impute the values for missing dates.

In [35]:
# Check for missing dates
missing_dates = date_range[~date_range.isin(client.date)]

# Print the missing dates
print("Missing Dates:")
print(missing_dates)

Missing Dates:
DatetimeIndex([], dtype='datetime64[ns]', freq='D')


* No missing dates in client data as well.

## 3.3 Explore weather_map data

In [36]:
# Shape
weather_map.shape

(112, 4)

In [37]:
# Check all columns and there types
weather_map.dtypes

county_name     object
longitude      float64
latitude       float64
county         float64
dtype: object

In [38]:
# Missing values
weather_map.isna().sum()

county_name    63
longitude       0
latitude        0
county         63
dtype: int64

In [39]:
weather_map[["county", "county_name"]].value_counts().sort_index()

county  county_name  
0.0     Harjumaa         6
1.0     Hiiumaa          1
2.0     Ida-Virumaa      4
3.0     Järvamaa         3
4.0     Jõgevamaa        3
5.0     Lääne-Virumaa    4
6.0     Läänemaa         2
7.0     Pärnumaa         5
8.0     Põlvamaa         1
9.0     Raplamaa         3
10.0    Saaremaa         4
11.0    Tartumaa         4
13.0    Valgamaa         1
14.0    Viljandimaa      3
15.0    Võrumaa          5
dtype: int64

In [40]:
weather_map.isna().sum()

county_name    63
longitude       0
latitude        0
county         63
dtype: int64

In [41]:
weather_map[weather_map.county.isna()]

county_name  longitude  latitude  county
0           NaN       21.7      57.6     NaN
1           NaN       21.7      57.9     NaN
2           NaN       21.7      58.2     NaN
3           NaN       21.7      58.5     NaN
4           NaN       21.7      58.8     NaN
..          ...        ...       ...     ...
107         NaN       28.2      58.5     NaN
108         NaN       28.2      58.8     NaN
109         NaN       28.2      59.1     NaN
110         NaN       28.2      59.4     NaN
111         NaN       28.2      59.7     NaN

[63 rows x 4 columns]

* County **12 - UNKOWN** is missing.

In [42]:
weather_map.county = weather_map.county.fillna(12)
weather_map.county_name = weather_map.county_name.fillna('UNKOWN')
weather_map.county = weather_map.county.astype("int")

In [43]:
weather_map.county.isna().sum()

0

## 3.4 Explore gas_price data

In [44]:
# Shape
gas_price.shape

(637, 5)

In [45]:
# Columns and data types
gas_price.dtypes

forecast_date            datetime64[ns]
lowest_price_per_mwh            float64
highest_price_per_mwh           float64
origin_date              datetime64[ns]
data_block_id                     int64
dtype: object

In [46]:
# Missing values
gas_price.isna().sum()

forecast_date            0
lowest_price_per_mwh     0
highest_price_per_mwh    0
origin_date              0
data_block_id            0
dtype: int64

In [47]:
# Dates for which gas price is available
start_date = gas_price.forecast_date.min()
end_date = gas_price.forecast_date.max()
print(start_date, end_date)

2021-09-01 00:00:00 2023-05-30 00:00:00


*  gas prices are not available for **2023-05-31**. WE will impute them using fill forward method.

## 3.5 Explore historic_weather data

In [48]:
# Shape
historic_weather.shape

(1710802, 18)

In [49]:
# Columns and datatypes
historic_weather.dtypes

datetime                  datetime64[ns]
temperature                      float64
dewpoint                         float64
rain                             float64
snowfall                         float64
surface_pressure                 float64
cloudcover_total                   int64
cloudcover_low                     int64
cloudcover_mid                     int64
cloudcover_high                    int64
windspeed_10m                    float64
winddirection_10m                  int64
shortwave_radiation              float64
direct_solar_radiation           float64
diffuse_radiation                float64
latitude                         float64
longitude                        float64
data_block_id                    float64
dtype: object

In [50]:
# Missing values
historic_weather.isna().sum()

datetime                  0
temperature               0
dewpoint                  0
rain                      0
snowfall                  0
surface_pressure          0
cloudcover_total          0
cloudcover_low            0
cloudcover_mid            0
cloudcover_high           0
windspeed_10m             0
winddirection_10m         0
shortwave_radiation       0
direct_solar_radiation    0
diffuse_radiation         0
latitude                  0
longitude                 0
data_block_id             0
dtype: int64

In [51]:
# Dates for which historic weather data is available
start_date = historic_weather.datetime.min()
end_date = historic_weather.datetime.max()
print(start_date, end_date)

2021-09-01 00:00:00 2023-05-30 10:00:00


*  historic weather is not available for **2023-05-31**. We can see if we can get that from the frecast data.

## 3.6 Explore forecast_weather data

In [52]:
# Shape
forecast_weather.shape

(3424512, 18)

In [53]:
# Columns and there data types
forecast_weather.dtypes

latitude                                    float64
longitude                                   float64
origin_datetime                      datetime64[ns]
hours_ahead                                   int64
temperature                                 float64
dewpoint                                    float64
cloudcover_high                             float64
cloudcover_low                              float64
cloudcover_mid                              float64
cloudcover_total                            float64
10_metre_u_wind_component                   float64
10_metre_v_wind_component                   float64
data_block_id                                 int64
forecast_datetime                    datetime64[ns]
direct_solar_radiation                      float64
surface_solar_radiation_downwards           float64
snowfall                                    float64
total_precipitation                         float64
dtype: object

In [54]:
# Missing values
forecast_weather.isna().sum()

latitude                             0
longitude                            0
origin_datetime                      0
hours_ahead                          0
temperature                          0
dewpoint                             0
cloudcover_high                      0
cloudcover_low                       0
cloudcover_mid                       0
cloudcover_total                     0
10_metre_u_wind_component            0
10_metre_v_wind_component            0
data_block_id                        0
forecast_datetime                    0
direct_solar_radiation               0
surface_solar_radiation_downwards    2
snowfall                             0
total_precipitation                  0
dtype: int64

* We have **2** missing values in **surface_solar_radiation_downwards**, let's check them in detail. 

In [55]:
forecast_weather[forecast_weather.surface_solar_radiation_downwards.isna()]

latitude  longitude     origin_datetime  hours_ahead  temperature  \
1849670      59.7       23.7 2022-08-11 02:00:00            3    19.043604   
1849782      59.7       23.7 2022-08-11 02:00:00            4    18.796777   

          dewpoint  cloudcover_high  cloudcover_low  cloudcover_mid  \
1849670  16.849023         0.908295             0.0             0.0   
1849782  16.994287         0.844788             0.0             0.0   

         cloudcover_total  10_metre_u_wind_component  \
1849670          0.908295                   5.913838   
1849782          0.844788                   5.421923   

         10_metre_v_wind_component  data_block_id   forecast_datetime  \
1849670                   7.620130            345 2022-08-11 05:00:00   
1849782                   8.103373            345 2022-08-11 06:00:00   

         direct_solar_radiation  surface_solar_radiation_downwards  snowfall  \
1849670               17.096667                                NaN       0.0   
1849782              206.413750                                NaN       0.0   

         total_precipitation  
1849670                  0.0  
1849782                  0.0

* These are morning hours **surface_solar_radiation_downwards** is 0, we will impute these values with 0.

In [56]:
forecast_weather[pd.to_datetime(forecast_weather["origin_datetime"].dt.date)==pd.to_datetime("2022-08-11")].sample(5)

latitude  longitude     origin_datetime  hours_ahead  temperature  \
1851379      57.9       24.2 2022-08-11 02:00:00           19    20.156763   
1854560      58.8       25.7 2022-08-11 02:00:00           47    14.322290   
1850120      59.7       24.7 2022-08-11 02:00:00            7    19.163477   
1853321      58.8       22.2 2022-08-11 02:00:00           36    19.394922   
1854071      57.9       26.2 2022-08-11 02:00:00           43    18.631250   

          dewpoint  cloudcover_high  cloudcover_low  cloudcover_mid  \
1851379  17.051904         0.234711             0.0             0.0   
1854560  13.013208         0.772888             0.0             0.0   
1850120  16.047388         0.995468             0.0             0.0   
1853321  17.360620         0.131012             0.0             0.0   
1854071  14.891260         0.356903             0.0             0.0   

         cloudcover_total  10_metre_u_wind_component  \
1851379          0.234711                   1.890010   
1854560          0.772888                   1.555817   
1850120          0.995468                   5.204838   
1853321          0.131012                   5.372705   
1854071          0.356903                   2.087479   

         10_metre_v_wind_component  data_block_id   forecast_datetime  \
1851379                   1.374841            345 2022-08-11 21:00:00   
1854560                   0.067767            345 2022-08-13 01:00:00   
1850120                   7.176234            345 2022-08-11 09:00:00   
1853321                   2.457233            345 2022-08-12 14:00:00   
1854071                   0.112079            345 2022-08-12 21:00:00   

         direct_solar_radiation  surface_solar_radiation_downwards  snowfall  \
1851379                3.422222                           0.586667       0.0   
1854560                0.000000                           0.000000       0.0   
1850120              649.882500                         406.272222       0.0   
1853321              842.773333                         683.244444       0.0   
1854071                0.013333                           0.000000       0.0   

         total_precipitation  
1851379                  0.0  
1854560                  0.0  
1850120                  0.0  
1853321                  0.0  
1854071                  0.0

In [57]:
forecast_weather.surface_solar_radiation_downwards.fillna(0, inplace=True)
# Missing values
forecast_weather.isna().sum()

latitude                             0
longitude                            0
origin_datetime                      0
hours_ahead                          0
temperature                          0
dewpoint                             0
cloudcover_high                      0
cloudcover_low                       0
cloudcover_mid                       0
cloudcover_total                     0
10_metre_u_wind_component            0
10_metre_v_wind_component            0
data_block_id                        0
forecast_datetime                    0
direct_solar_radiation               0
surface_solar_radiation_downwards    0
snowfall                             0
total_precipitation                  0
dtype: int64

* Good, no missing values now.

In [58]:
# Dates for which historic weather data is available
start_date = forecast_weather.forecast_datetime.min()
end_date = forecast_weather.forecast_datetime.max()
print(start_date, end_date)

2021-09-01 03:00:00 2023-06-01 02:00:00


* We have forecast data for next week from the last date in train data till **01-06-2023**

# 4. Merging Data

## 4.1 Merging client data

In [59]:
train.sample(5)

county  is_business  product_type  is_consumption  \
864330       11            1             1               0   
828111        7            0             3               1   
230799        9            0             3               1   
1884152      14            0             3               0   
517930       14            1             1               0   

                   datetime  data_block_id   row_id  prediction_unit_id  \
864330  2022-06-06 01:00:00            278   864330                  46   
828111  2022-05-25 19:00:00            266   828111                  27   
230799  2021-11-17 06:00:00             77   230799                  35   
1884152 2023-04-19 21:00:00            595  1884152                  54   
517930  2022-02-18 09:00:00            170   517930                  55   

              date      time   target  
864330  2022-06-06  01:00:00    0.000  
828111  2022-05-25  19:00:00   98.463  
230799  2021-11-17  06:00:00  134.213  
1884152 2023-04-19  21:00:00    2.787  
517930  2022-02-18  09:00:00    1.879

In [60]:
client.sort_values(by=["county", "is_business", "product_type", "date"]).tail(5)

product_type  county  eic_count  installed_capacity  is_business  \
41656             3      15         55              2188.2            1   
41722             3      15         55              2188.2            1   
41788             3      15         55              2188.2            1   
41853             3      15         55              2188.2            1   
41918             3      15         55              2188.2            1   

            date  data_block_id  
41656 2023-05-25            633  
41722 2023-05-26            634  
41788 2023-05-27            635  
41853 2023-05-28            636  
41918 2023-05-29            637

In [61]:
train.shape, client.shape

((2018352, 11), (41919, 7))

In [62]:
train_m = train.merge(
    client,
    how="left",
    on=["county", "is_business", "product_type", "date"],
    suffixes=("", "_client")
) 
train_m.shape

(2018352, 14)

In [63]:
del train

In [64]:
train_m.isna().sum()

county                     0
is_business                0
product_type               0
is_consumption             0
datetime                   0
data_block_id              0
row_id                     0
prediction_unit_id         0
date                       0
time                       0
target                     0
eic_count               6240
installed_capacity      6240
data_block_id_client    6240
dtype: int64

* We have missing data for some dates in client data, we will impute it with previous dates.

In [65]:
train_m[["county", "is_business", "product_type", "date", "eic_count", "installed_capacity"]]

county  is_business  product_type       date  eic_count  \
0             0            0             1 2021-09-01      108.0   
1             0            0             1 2021-09-01      108.0   
2             0            0             2 2021-09-01       17.0   
3             0            0             2 2021-09-01       17.0   
4             0            0             3 2021-09-01      688.0   
...         ...          ...           ...        ...        ...   
2018347      15            1             0 2023-05-31        NaN   
2018348      15            1             1 2023-05-31        NaN   
2018349      15            1             1 2023-05-31        NaN   
2018350      15            1             3 2023-05-31        NaN   
2018351      15            1             3 2023-05-31        NaN   

         installed_capacity  
0                    952.89  
1                    952.89  
2                    166.40  
3                    166.40  
4                   7207.88  
...                     ...  
2018347                 NaN  
2018348                 NaN  
2018349                 NaN  
2018350                 NaN  
2018351                 NaN  

[2018352 rows x 6 columns]

In [66]:
missing_client = train_m.groupby(["county", "is_business", "product_type", "date"])["eic_count", "installed_capacity"].mean().reset_index().ffill()

In [67]:
missing_client.tail()

county  is_business  product_type       date  eic_count  \
42044      15            1             3 2023-05-27       55.0   
42045      15            1             3 2023-05-28       55.0   
42046      15            1             3 2023-05-29       55.0   
42047      15            1             3 2023-05-30       55.0   
42048      15            1             3 2023-05-31       55.0   

       installed_capacity  
42044              2188.2  
42045              2188.2  
42046              2188.2  
42047              2188.2  
42048              2188.2

In [68]:
train_m.drop(columns=["eic_count", "installed_capacity", "data_block_id_client"], inplace=True)
train_m = train_m.merge(
    missing_client,
    on=["county", "is_business", "product_type", "date"],
    suffixes=("", "_client")
)
train_m.shape

(2018352, 13)

In [69]:
train_m.isna().sum()

county                0
is_business           0
product_type          0
is_consumption        0
datetime              0
data_block_id         0
row_id                0
prediction_unit_id    0
date                  0
time                  0
target                0
eic_count             0
installed_capacity    0
dtype: int64

* No missing data anymore, let's move to next dataset.

## 4.2 Merging weather_map with forecast_weather data

In [70]:
weather_map.sample(5)

county_name  longitude  latitude  county
36       Läänemaa       23.7      58.8       6
61       Harjumaa       25.2      59.1       0
34         UNKOWN       23.7      58.2      12
41         UNKOWN       24.2      57.9      12
85  Lääne-Virumaa       26.7      59.1       5

In [71]:
forecast_weather.sample(2)

latitude  longitude     origin_datetime  hours_ahead  temperature  \
975593       59.1       23.2 2022-03-01 01:00:00           23     0.905176   
2048676      59.4       21.7 2022-09-17 02:00:00            4    12.388696   

         dewpoint  cloudcover_high  cloudcover_low  cloudcover_mid  \
975593  -0.694678         0.959442        0.000000        0.000000   
2048676  9.097925         0.377411        0.995979        0.837708   

         cloudcover_total  10_metre_u_wind_component  \
975593           0.959442                   4.224582   
2048676          0.999328                   3.622288   

         10_metre_v_wind_component  data_block_id   forecast_datetime  \
975593                    4.492805            182 2022-03-02 00:00:00   
2048676                  -1.442930            382 2022-09-17 06:00:00   

         direct_solar_radiation  surface_solar_radiation_downwards  snowfall  \
975593                      0.0                                0.0       0.0   
2048676                     0.0                                0.0       0.0   

         total_precipitation  
975593              0.000000  
2048676             0.001238

In [72]:
forecast_weather.shape

(3424512, 18)

In [73]:
weather_map.longitude.dtype, weather_map.latitude.dtype, forecast_weather.longitude.dtype, forecast_weather.latitude.dtype

(dtype('float64'), dtype('float64'), dtype('float64'), dtype('float64'))

In [74]:
weather_map.latitude = weather_map.latitude.round(1)
weather_map.longitude = weather_map.longitude.round(1)
forecast_weather.latitude = forecast_weather.latitude.round(1)
forecast_weather.longitude = forecast_weather.longitude.round(1)

In [75]:
forecast_weather_m = forecast_weather.merge(
    weather_map,
    how="left",
    on=["longitude", "latitude"],
    suffixes=("", "_map")
)
forecast_weather_m.shape

(3424512, 20)

In [76]:
forecast_weather_m.isna().sum()

latitude                             0
longitude                            0
origin_datetime                      0
hours_ahead                          0
temperature                          0
dewpoint                             0
cloudcover_high                      0
cloudcover_low                       0
cloudcover_mid                       0
cloudcover_total                     0
10_metre_u_wind_component            0
10_metre_v_wind_component            0
data_block_id                        0
forecast_datetime                    0
direct_solar_radiation               0
surface_solar_radiation_downwards    0
snowfall                             0
total_precipitation                  0
county_name                          0
county                               0
dtype: int64

In [77]:
forecast_weather_m.county.value_counts()

12    1926288
0      183456
15     152880
7      152880
10     122304
11     122304
5      122304
2      122304
14      91728
4       91728
9       91728
3       91728
6       61152
13      30576
8       30576
1       30576
Name: county, dtype: int64

In [78]:
del forecast_weather

## 4.3 Merging weather_map with historic_weather data

In [79]:
weather_map.sample(5)

county_name  longitude  latitude  county
80       Võrumaa       26.7      57.6      15
68      Järvamaa       25.7      58.8       3
2         UNKOWN       21.7      58.2      12
21        UNKOWN       22.7      59.1      12
102  Ida-Virumaa       27.7      59.4       2

In [80]:
historic_weather.sample(2)

datetime  temperature  dewpoint  rain  snowfall  \
678811 2022-05-11 12:00:00         10.8       8.2   0.0       0.0   
241669 2021-11-29 21:00:00         -1.2      -4.6   0.0       0.0   

        surface_pressure  cloudcover_total  cloudcover_low  cloudcover_mid  \
678811            1001.0                94             100               6   
241669             997.8               100              99              96   

        cloudcover_high  windspeed_10m  winddirection_10m  \
678811                0       6.444444                228   
241669               99       8.861111                 27   

        shortwave_radiation  direct_solar_radiation  diffuse_radiation  \
678811                217.0                    16.0              201.0   
241669                  0.0                     0.0                0.0   

        latitude  longitude  data_block_id  
678811      59.4       25.2          254.0  
241669      59.4       22.2           91.0

In [81]:
historic_weather.shape

(1710802, 18)

In [82]:
historic_weather.latitude = historic_weather.latitude.round(1)
historic_weather.longitude = historic_weather.longitude.round(1)

In [83]:
historic_weather_m = historic_weather.merge(
    weather_map,
    how="left",
    on=["latitude", "longitude"],
    suffixes=("", "_map")
)
historic_weather_m.shape

(1710802, 20)

In [84]:
historic_weather_m.isna().sum()

datetime                  0
temperature               0
dewpoint                  0
rain                      0
snowfall                  0
surface_pressure          0
cloudcover_total          0
cloudcover_low            0
cloudcover_mid            0
cloudcover_high           0
windspeed_10m             0
winddirection_10m         0
shortwave_radiation       0
direct_solar_radiation    0
diffuse_radiation         0
latitude                  0
longitude                 0
data_block_id             0
county_name               0
county                    0
dtype: int64

* Great!! No missing values.

In [85]:
del historic_weather

## 4.4 Merging gas prices to train data

In [86]:
gas_price.tail(5)

forecast_date  lowest_price_per_mwh  highest_price_per_mwh origin_date  \
632    2023-05-26                 29.10                  34.10  2023-05-25   
633    2023-05-27                 28.30                  34.10  2023-05-26   
634    2023-05-28                 28.10                  34.10  2023-05-27   
635    2023-05-29                 28.16                  36.98  2023-05-28   
636    2023-05-30                 29.00                  34.00  2023-05-29   

     data_block_id  
632            633  
633            634  
634            635  
635            636  
636            637

In [87]:
train_m.tail(2)

county  is_business  product_type  is_consumption  \
2018350      15            1             3               0   
2018351      15            1             3               1   

                   datetime  data_block_id   row_id  prediction_unit_id  \
2018350 2023-05-31 23:00:00            637  2018350                  60   
2018351 2023-05-31 23:00:00            637  2018351                  60   

              date      time  target  eic_count  installed_capacity  
2018350 2023-05-31  23:00:00    0.00       55.0              2188.2  
2018351 2023-05-31  23:00:00  196.24       55.0              2188.2

In [88]:
train_m.shape

(2018352, 13)

In [89]:
gas_price.forecast_date.dtype, train_m.date.dtype

(dtype('<M8[ns]'), dtype('<M8[ns]'))

In [90]:
gas_price.rename(columns={"forecast_date": "date"}, inplace=True)

In [91]:
train_m_2 = train_m.merge(
    gas_price,
    how="left",
    on="date",
    suffixes=("", "_gas")
)
train_m_2.shape

(2018352, 17)

In [92]:
train_m_2.isna().sum()

county                      0
is_business                 0
product_type                0
is_consumption              0
datetime                    0
data_block_id               0
row_id                      0
prediction_unit_id          0
date                        0
time                        0
target                      0
eic_count                   0
installed_capacity          0
lowest_price_per_mwh     3120
highest_price_per_mwh    3120
origin_date              3120
data_block_id_gas        3120
dtype: int64

In [93]:
train_m_2[train_m_2.lowest_price_per_mwh.isna()]["date"].value_counts()

2023-05-31    3120
Name: date, dtype: int64

* We do not have gas price data for *31*, let's imput this with previous day values.

In [94]:
train_m_2[["county", "is_business", "product_type", "date", "lowest_price_per_mwh", "highest_price_per_mwh", "origin_date", "data_block_id_gas"]]

county  is_business  product_type       date  lowest_price_per_mwh  \
0             0            0             1 2021-09-01                 45.23   
1             0            0             1 2021-09-01                 45.23   
2             0            0             1 2021-09-01                 45.23   
3             0            0             1 2021-09-01                 45.23   
4             0            0             1 2021-09-01                 45.23   
...         ...          ...           ...        ...                   ...   
2018347      15            1             3 2023-05-31                   NaN   
2018348      15            1             3 2023-05-31                   NaN   
2018349      15            1             3 2023-05-31                   NaN   
2018350      15            1             3 2023-05-31                   NaN   
2018351      15            1             3 2023-05-31                   NaN   

         highest_price_per_mwh origin_date  data_block_id_gas  
0                        46.32  2021-08-31                1.0  
1                        46.32  2021-08-31                1.0  
2                        46.32  2021-08-31                1.0  
3                        46.32  2021-08-31                1.0  
4                        46.32  2021-08-31                1.0  
...                        ...         ...                ...  
2018347                    NaN         NaT                NaN  
2018348                    NaN         NaT                NaN  
2018349                    NaN         NaT                NaN  
2018350                    NaN         NaT                NaN  
2018351                    NaN         NaT                NaN  

[2018352 rows x 8 columns]

In [95]:
missing_gas = train_m_2.groupby(["county", "is_business", "product_type", "date"])["lowest_price_per_mwh", "highest_price_per_mwh"].mean().reset_index().ffill()

In [96]:
missing_gas.tail()

county  is_business  product_type       date  lowest_price_per_mwh  \
42044      15            1             3 2023-05-27                 28.30   
42045      15            1             3 2023-05-28                 28.10   
42046      15            1             3 2023-05-29                 28.16   
42047      15            1             3 2023-05-30                 29.00   
42048      15            1             3 2023-05-31                 29.00   

       highest_price_per_mwh  
42044                  34.10  
42045                  34.10  
42046                  36.98  
42047                  34.00  
42048                  34.00

In [97]:
train_m_2.drop(columns=["lowest_price_per_mwh", "highest_price_per_mwh", "origin_date", "data_block_id_gas"], inplace=True)
train_m_2 = train_m_2.merge(
    missing_gas,
    how="left",
    on=["county", "is_business", "product_type", "date"],
    suffixes=("", "_gas")
)
train_m_2.shape

(2018352, 15)

In [98]:
train_m_2.isna().sum()

county                   0
is_business              0
product_type             0
is_consumption           0
datetime                 0
data_block_id            0
row_id                   0
prediction_unit_id       0
date                     0
time                     0
target                   0
eic_count                0
installed_capacity       0
lowest_price_per_mwh     0
highest_price_per_mwh    0
dtype: int64

* No missing data anymore, let's move to next dataset.

In [99]:
del train_m

## 4.5 Merging electricity prices to train data

In [100]:
electricity_price.tail(5)

forecast_date  euros_per_mwh         origin_date  data_block_id
15281 2023-05-30 19:00:00          82.10 2023-05-29 19:00:00            637
15282 2023-05-30 20:00:00         150.85 2023-05-29 20:00:00            637
15283 2023-05-30 21:00:00          82.10 2023-05-29 21:00:00            637
15284 2023-05-30 22:00:00          82.09 2023-05-29 22:00:00            637
15285 2023-05-30 23:00:00          -1.29 2023-05-29 23:00:00            637

In [101]:
train_m_2.tail(2)

county  is_business  product_type  is_consumption  \
2018350      15            1             3               0   
2018351      15            1             3               1   

                   datetime  data_block_id   row_id  prediction_unit_id  \
2018350 2023-05-31 23:00:00            637  2018350                  60   
2018351 2023-05-31 23:00:00            637  2018351                  60   

              date      time  target  eic_count  installed_capacity  \
2018350 2023-05-31  23:00:00    0.00       55.0              2188.2   
2018351 2023-05-31  23:00:00  196.24       55.0              2188.2   

         lowest_price_per_mwh  highest_price_per_mwh  
2018350                  29.0                   34.0  
2018351                  29.0                   34.0

In [102]:
train_m_2.shape

(2018352, 15)

In [103]:
electricity_price.forecast_date.dtype, train_m_2.datetime.dtype

(dtype('<M8[ns]'), dtype('<M8[ns]'))

In [104]:
electricity_price.rename(columns={"forecast_date": "datetime"}, inplace=True)

In [105]:
train_m_3 = train_m_2.merge(
    electricity_price,
    how="left",
    on="datetime",
    suffixes=("", "_elec")
)
train_m_3.shape

(2018352, 18)

In [106]:
train_m_3.isna().sum()

county                      0
is_business                 0
product_type                0
is_consumption              0
datetime                    0
data_block_id               0
row_id                      0
prediction_unit_id          0
date                        0
time                        0
target                      0
eic_count                   0
installed_capacity          0
lowest_price_per_mwh        0
highest_price_per_mwh       0
euros_per_mwh            3386
origin_date              3386
data_block_id_elec       3386
dtype: int64

In [107]:
train_m_3[train_m_3.euros_per_mwh.isna()]["date"].value_counts()

2023-05-31    3120
2022-03-27     134
2023-03-26     132
Name: date, dtype: int64

* We do not have electricity price data for *31* and for hour 2:00:00 for two other dates. let's imput this with previous day values.

In [108]:
train_m_3[["county", "is_business", "product_type", "datetime", "euros_per_mwh", "origin_date", "data_block_id_elec"]]

county  is_business  product_type            datetime  euros_per_mwh  \
0             0            0             1 2021-09-01 00:00:00          92.51   
1             0            0             1 2021-09-01 00:00:00          92.51   
2             0            0             1 2021-09-01 01:00:00          88.90   
3             0            0             1 2021-09-01 01:00:00          88.90   
4             0            0             1 2021-09-01 02:00:00          87.35   
...         ...          ...           ...                 ...            ...   
2018347      15            1             3 2023-05-31 21:00:00            NaN   
2018348      15            1             3 2023-05-31 22:00:00            NaN   
2018349      15            1             3 2023-05-31 22:00:00            NaN   
2018350      15            1             3 2023-05-31 23:00:00            NaN   
2018351      15            1             3 2023-05-31 23:00:00            NaN   

                origin_date  data_block_id_elec  
0       2021-08-31 00:00:00                 1.0  
1       2021-08-31 00:00:00                 1.0  
2       2021-08-31 01:00:00                 1.0  
3       2021-08-31 01:00:00                 1.0  
4       2021-08-31 02:00:00                 1.0  
...                     ...                 ...  
2018347                 NaT                 NaN  
2018348                 NaT                 NaN  
2018349                 NaT                 NaN  
2018350                 NaT                 NaN  
2018351                 NaT                 NaN  

[2018352 rows x 7 columns]

In [109]:
train_m_3["time"] = train_m_3.datetime.dt.time

In [110]:
train_m_3[["county", "is_business", "product_type", "is_consumption", "time", "date"]].duplicated().value_counts()

False    2018352
dtype: int64

In [111]:
missing_elec = train_m_3.groupby(["county", "is_business", "product_type", "is_consumption", "time", "date"])["euros_per_mwh"].mean().reset_index().ffill()

In [112]:
train_m_3[(train_m_3.date == pd.to_datetime("2023-05-29")) & (train_m_3.county == 15) & (train_m_3.is_business == 1) & (train_m_3.product_type == 3)].tail(5)

county  is_business  product_type  is_consumption  \
2012107      15            1             3               1   
2012108      15            1             3               0   
2012109      15            1             3               1   
2012110      15            1             3               0   
2012111      15            1             3               1   

                   datetime  data_block_id   row_id  prediction_unit_id  \
2012107 2023-05-29 21:00:00            635  2011851                  60   
2012108 2023-05-29 22:00:00            635  2011980                  60   
2012109 2023-05-29 22:00:00            635  2011981                  60   
2012110 2023-05-29 23:00:00            635  2012110                  60   
2012111 2023-05-29 23:00:00            635  2012111                  60   

              date      time   target  eic_count  installed_capacity  \
2012107 2023-05-29  21:00:00  296.073       55.0              2188.2   
2012108 2023-05-29  22:00:00    0.000       55.0              2188.2   
2012109 2023-05-29  22:00:00  299.806       55.0              2188.2   
2012110 2023-05-29  23:00:00    0.000       55.0              2188.2   
2012111 2023-05-29  23:00:00  177.056       55.0              2188.2   

         lowest_price_per_mwh  highest_price_per_mwh  euros_per_mwh  \
2012107                 28.16                  36.98          82.30   
2012108                 28.16                  36.98          82.29   
2012109                 28.16                  36.98          82.29   
2012110                 28.16                  36.98          77.27   
2012111                 28.16                  36.98          77.27   

                origin_date  data_block_id_elec  
2012107 2023-05-28 21:00:00               636.0  
2012108 2023-05-28 22:00:00               636.0  
2012109 2023-05-28 22:00:00               636.0  
2012110 2023-05-28 23:00:00               636.0  
2012111 2023-05-28 23:00:00               636.0

In [113]:
missing_elec[(missing_elec.date == pd.to_datetime("2023-05-29"))].tail(5)

county  is_business  product_type  is_consumption      time  \
2015797      15            1             3               1  19:00:00   
2016435      15            1             3               1  20:00:00   
2017073      15            1             3               1  21:00:00   
2017711      15            1             3               1  22:00:00   
2018349      15            1             3               1  23:00:00   

              date  euros_per_mwh  
2015797 2023-05-29          82.30  
2016435 2023-05-29          82.30  
2017073 2023-05-29          82.30  
2017711 2023-05-29          82.29  
2018349 2023-05-29          77.27

In [114]:
missing_elec.tail()

county  is_business  product_type  is_consumption      time  \
2018347      15            1             3               1  23:00:00   
2018348      15            1             3               1  23:00:00   
2018349      15            1             3               1  23:00:00   
2018350      15            1             3               1  23:00:00   
2018351      15            1             3               1  23:00:00   

              date  euros_per_mwh  
2018347 2023-05-27          94.71  
2018348 2023-05-28          69.50  
2018349 2023-05-29          77.27  
2018350 2023-05-30          -1.29  
2018351 2023-05-31          -1.29

In [115]:
train_m_3.drop(columns=["euros_per_mwh", "origin_date", "data_block_id_elec"], inplace=True)
train_m_3 = train_m_3.merge(
    missing_elec,
    how="left",
    on=["county", "is_business", "product_type", "is_consumption", "time", "date"],
    suffixes=("", "_elec")
)
train_m_3.shape

(2018352, 16)

In [116]:
train_m_3.isna().sum()

county                   0
is_business              0
product_type             0
is_consumption           0
datetime                 0
data_block_id            0
row_id                   0
prediction_unit_id       0
date                     0
time                     0
target                   0
eic_count                0
installed_capacity       0
lowest_price_per_mwh     0
highest_price_per_mwh    0
euros_per_mwh            0
dtype: int64

* No missing data anymore, let's move to next dataset.

In [117]:
del train_m_2

## 4.6 Merging forecast_weather to train data

In [118]:
forecast_weather_m.head(3)

latitude  longitude     origin_datetime  hours_ahead  temperature  \
0      57.6       21.7 2021-09-01 02:00:00            1    15.655786   
1      57.6       22.2 2021-09-01 02:00:00            1    13.003931   
2      57.6       22.7 2021-09-01 02:00:00            1    14.206567   

    dewpoint  cloudcover_high  cloudcover_low  cloudcover_mid  \
0  11.553613         0.904816        0.019714             0.0   
1  10.689844         0.886322        0.004456             0.0   
2  11.671777         0.729034        0.005615             0.0   

   cloudcover_total  10_metre_u_wind_component  10_metre_v_wind_component  \
0          0.905899                  -0.411328                  -9.106137   
1          0.886658                   0.206347                  -5.355405   
2          0.730499                   1.451587                  -7.417905   

   data_block_id   forecast_datetime  direct_solar_radiation  \
0              1 2021-09-01 03:00:00                     0.0   
1              1 2021-09-01 03:00:00                     0.0   
2              1 2021-09-01 03:00:00                     0.0   

   surface_solar_radiation_downwards  snowfall  total_precipitation  \
0                                0.0       0.0                  0.0   
1                                0.0       0.0                  0.0   
2                                0.0       0.0                  0.0   

  county_name  county  
0      UNKOWN      12  
1      UNKOWN      12  
2      UNKOWN      12

In [119]:
train_m_3.tail(2)

county  is_business  product_type  is_consumption  \
2018350      15            1             3               0   
2018351      15            1             3               1   

                   datetime  data_block_id   row_id  prediction_unit_id  \
2018350 2023-05-31 23:00:00            637  2018350                  60   
2018351 2023-05-31 23:00:00            637  2018351                  60   

              date      time  target  eic_count  installed_capacity  \
2018350 2023-05-31  23:00:00    0.00       55.0              2188.2   
2018351 2023-05-31  23:00:00  196.24       55.0              2188.2   

         lowest_price_per_mwh  highest_price_per_mwh  euros_per_mwh  
2018350                  29.0                   34.0          -1.29  
2018351                  29.0                   34.0          -1.29

In [120]:
# forecast_weather_m.hours_ahead.value_counts()

In [121]:
forecast_weather_m.forecast_datetime.dtype, train_m_3.datetime.dtype

(dtype('<M8[ns]'), dtype('<M8[ns]'))

In [122]:
forecast_weather_m.rename(columns={"forecast_datetime": "datetime"}, inplace=True)

In [123]:
forecast_weather_m = forecast_weather_m.groupby(["county", "datetime"]).mean().reset_index()

In [124]:
forecast_weather_m["county"] = forecast_weather_m.county.astype("int")

In [125]:
train_m_3.shape

(2018352, 16)

In [126]:
train_m_4 = train_m_3.merge(
    forecast_weather_m,
    how="left",
    on=["county", "datetime"],
    suffixes=("", "_fw") 
)
train_m_4.shape

(2018352, 32)

In [127]:
train_m_4.isna().sum()

county                                 0
is_business                            0
product_type                           0
is_consumption                         0
datetime                               0
data_block_id                          0
row_id                                 0
prediction_unit_id                     0
date                                   0
time                                   0
target                                 0
eic_count                              0
installed_capacity                     0
lowest_price_per_mwh                   0
highest_price_per_mwh                  0
euros_per_mwh                          0
latitude                             632
longitude                            632
hours_ahead                          632
temperature                          632
dewpoint                             632
cloudcover_high                      632
cloudcover_low                       632
cloudcover_mid                       632
cloudcover_total

In [128]:
train_m_4[train_m_4.hours_ahead.isna()].datetime.value_counts()

2022-03-27 03:00:00    134
2023-03-26 03:00:00    132
2021-09-01 00:00:00    122
2021-09-01 01:00:00    122
2021-09-01 02:00:00    122
Name: datetime, dtype: int64

* We do not have forecast weather data for some dates. let's imput this with previous day values.

In [129]:
train_m_4[["county", "is_business", "product_type", "datetime", "temperature"]]

county  is_business  product_type            datetime  temperature
0             0            0             1 2021-09-01 00:00:00          NaN
1             0            0             1 2021-09-01 00:00:00          NaN
2             0            0             1 2021-09-01 01:00:00          NaN
3             0            0             1 2021-09-01 01:00:00          NaN
4             0            0             1 2021-09-01 02:00:00          NaN
...         ...          ...           ...                 ...          ...
2018347      15            1             3 2023-05-31 21:00:00    13.916846
2018348      15            1             3 2023-05-31 22:00:00    12.470825
2018349      15            1             3 2023-05-31 22:00:00    12.470825
2018350      15            1             3 2023-05-31 23:00:00    11.484033
2018351      15            1             3 2023-05-31 23:00:00    11.484033

[2018352 rows x 5 columns]

In [130]:
train_m_4[["county", "is_business", "product_type", "is_consumption", "time", "date"]].duplicated().value_counts()

False    2018352
dtype: int64

In [131]:
missing_fw = train_m_4.groupby(["county", "is_business", "product_type", "is_consumption", "time", "date"])["temperature","dewpoint","cloudcover_high","cloudcover_low","cloudcover_mid","cloudcover_total","10_metre_u_wind_component","10_metre_v_wind_component","data_block_id_fw","direct_solar_radiation","surface_solar_radiation_downwards","snowfall","total_precipitation"].mean().reset_index().ffill()
missing_fw = missing_fw.bfill()

In [132]:
train_m_4[(train_m_4.date == pd.to_datetime("2023-03-26")) & (train_m_4.county == 15) & (train_m_4.is_business == 1) & (train_m_4.product_type == 3)].head(8)

county  is_business  product_type  is_consumption  \
1808976      15            1             3               0   
1808977      15            1             3               1   
1808978      15            1             3               0   
1808979      15            1             3               1   
1808980      15            1             3               0   
1808981      15            1             3               1   
1808982      15            1             3               0   
1808983      15            1             3               1   

                   datetime  data_block_id   row_id  prediction_unit_id  \
1808976 2023-03-26 00:00:00            571  1805986                  60   
1808977 2023-03-26 00:00:00            571  1805987                  60   
1808978 2023-03-26 01:00:00            571  1806118                  60   
1808979 2023-03-26 01:00:00            571  1806119                  60   
1808980 2023-03-26 02:00:00            571  1806250                  60   
1808981 2023-03-26 02:00:00            571  1806251                  60   
1808982 2023-03-26 03:00:00            571  1806382                  60   
1808983 2023-03-26 03:00:00            571  1806383                  60   

              date      time  ...  cloudcover_low  cloudcover_mid  \
1808976 2023-03-26  00:00:00  ...        0.307510        0.457578   
1808977 2023-03-26  00:00:00  ...        0.307510        0.457578   
1808978 2023-03-26  01:00:00  ...        0.549413        0.142081   
1808979 2023-03-26  01:00:00  ...        0.549413        0.142081   
1808980 2023-03-26  02:00:00  ...        0.541530        0.122595   
1808981 2023-03-26  02:00:00  ...        0.541530        0.122595   
1808982 2023-03-26  03:00:00  ...             NaN             NaN   
1808983 2023-03-26  03:00:00  ...             NaN             NaN   

         cloudcover_total  10_metre_u_wind_component  \
1808976          0.902218                   1.449661   
1808977          0.902218                   1.449661   
1808978          0.966574                   1.487020   
1808979          0.966574                   1.487020   
1808980          0.661326                   1.555820   
1808981          0.661326                   1.555820   
1808982               NaN                        NaN   
1808983               NaN                        NaN   

         10_metre_v_wind_component  data_block_id_fw  direct_solar_radiation  \
1808976                   1.515520             570.5                     0.0   
1808977                   1.515520             570.5                     0.0   
1808978                   1.497897             570.5                     0.0   
1808979                   1.497897             570.5                     0.0   
1808980                   1.498167             571.5                     0.0   
1808981                   1.498167             571.5                     0.0   
1808982                        NaN               NaN                     NaN   
1808983                        NaN               NaN                     NaN   

         surface_solar_radiation_downwards  snowfall  total_precipitation  
1808976                                0.0  0.000001             0.000239  
1808977                                0.0  0.000001             0.000239  
1808978                                0.0  0.000000             0.000028  
1808979                                0.0  0.000000             0.000028  
1808980                                0.0  0.000000             0.000003  
1808981                                0.0  0.000000             0.000003  
1808982                                NaN       NaN                  NaN  
1808983                                NaN       NaN                  NaN  

[8 rows x 32 columns]

In [133]:
missing_fw[(missing_fw.date == pd.to_datetime("2023-03-26"))].head()

county  is_business  product_type  is_consumption      time       date  \
571        0            0             1               0  00:00:00 2023-03-26   
1209       0            0             1               0  01:00:00 2023-03-26   
1847       0            0             1               0  02:00:00 2023-03-26   
2485       0            0             1               0  03:00:00 2023-03-26   
3123       0            0             1               0  04:00:00 2023-03-26   

      temperature  dewpoint  cloudcover_high  cloudcover_low  cloudcover_mid  \
571      3.764520  3.160537         0.664223        1.000000        0.597473   
1209     3.765833  3.277256         0.557884        1.000001        0.404747   
1847     3.784113  3.506982         0.094894        0.999999        0.160533   
2485     5.226272  3.108118         0.143425        1.000001        0.295176   
3123     3.662602  3.446029         0.000389        0.999999        0.081057   

      cloudcover_total  10_metre_u_wind_component  10_metre_v_wind_component  \
571           1.000001                   0.614123                   2.390402   
1209          1.000006                   0.988135                   2.054371   
1847          0.999997                   1.031340                   0.896555   
2485          1.000001                   4.941571                   6.262753   
3123          1.000000                   0.560785                   0.736835   

      data_block_id_fw  direct_solar_radiation  \
571              570.5                     0.0   
1209             570.5                     0.0   
1847             571.5                     0.0   
2485             570.5                     0.0   
3123             571.5                     0.0   

      surface_solar_radiation_downwards  snowfall  total_precipitation  
571                            0.000000       0.0             0.000131  
1209                           0.000000       0.0             0.000126  
1847                           0.000000       0.0             0.000043  
2485                          -0.019028       0.0             0.000002  
3123                          -0.016759       0.0             0.000035

In [134]:
train_m_4.drop(columns=["latitude","longitude","hours_ahead","temperature","dewpoint","cloudcover_high","cloudcover_low","cloudcover_mid","cloudcover_total","10_metre_u_wind_component","10_metre_v_wind_component","data_block_id_fw","direct_solar_radiation","surface_solar_radiation_downwards","snowfall","total_precipitation"], inplace=True)
train_m_4 = train_m_4.merge(
    missing_fw,
    how="left",
    on=["county", "is_business", "product_type", "is_consumption", "time", "date"],
    suffixes=("", "_fw")
)
train_m_4.shape

(2018352, 29)

In [135]:
train_m_4.isna().sum()

county                               0
is_business                          0
product_type                         0
is_consumption                       0
datetime                             0
data_block_id                        0
row_id                               0
prediction_unit_id                   0
date                                 0
time                                 0
target                               0
eic_count                            0
installed_capacity                   0
lowest_price_per_mwh                 0
highest_price_per_mwh                0
euros_per_mwh                        0
temperature                          0
dewpoint                             0
cloudcover_high                      0
cloudcover_low                       0
cloudcover_mid                       0
cloudcover_total                     0
10_metre_u_wind_component            0
10_metre_v_wind_component            0
data_block_id_fw                     0
direct_solar_radiation   

* No missing data anymore, let's move to next dataset.

In [136]:
del train_m_3

## 4.7 Merging historic_weather to train data

In [137]:
historic_weather_m.head(3)

datetime  temperature  dewpoint  rain  snowfall  surface_pressure  \
0 2021-09-01         14.2      11.6   0.0       0.0            1015.9   
1 2021-09-01         13.9      11.5   0.0       0.0            1010.7   
2 2021-09-01         14.0      12.5   0.0       0.0            1015.0   

   cloudcover_total  cloudcover_low  cloudcover_mid  cloudcover_high  \
0                31              31               0               11   
1                33              37               0                0   
2                31              34               0                0   

   windspeed_10m  winddirection_10m  shortwave_radiation  \
0       7.083333                  8                  0.0   
1       5.111111                359                  0.0   
2       6.333333                355                  0.0   

   direct_solar_radiation  diffuse_radiation  latitude  longitude  \
0                     0.0                0.0      57.6       21.7   
1                     0.0                0.0      57.6       22.2   
2                     0.0                0.0      57.6       22.7   

   data_block_id county_name  county  
0            1.0      UNKOWN      12  
1            1.0      UNKOWN      12  
2            1.0      UNKOWN      12

In [138]:
train_m_4.tail(2)

county  is_business  product_type  is_consumption  \
2018350      15            1             3               0   
2018351      15            1             3               1   

                   datetime  data_block_id   row_id  prediction_unit_id  \
2018350 2023-05-31 23:00:00            637  2018350                  60   
2018351 2023-05-31 23:00:00            637  2018351                  60   

              date      time  ...  cloudcover_low  cloudcover_mid  \
2018350 2023-05-31  23:00:00  ...        0.045816        0.017661   
2018351 2023-05-31  23:00:00  ...        0.045816        0.017661   

         cloudcover_total  10_metre_u_wind_component  \
2018350          0.488751                   3.560166   
2018351          0.488751                   3.560166   

         10_metre_v_wind_component  data_block_id_fw  direct_solar_radiation  \
2018350                  -0.286433             637.0               -0.335111   
2018351                  -0.286433             637.0               -0.335111   

         surface_solar_radiation_downwards  snowfall  total_precipitation  
2018350                                0.0       0.0                  0.0  
2018351                                0.0       0.0                  0.0  

[2 rows x 29 columns]

In [139]:
historic_weather_m.datetime.dtype, train_m_4.datetime.dtype

(dtype('<M8[ns]'), dtype('<M8[ns]'))

In [140]:
historic_weather_m = historic_weather_m.groupby(["county", "datetime"]).mean().reset_index()

In [141]:
historic_weather_m["county"] = historic_weather_m.county.astype("int")

In [142]:
train_m_4.shape

(2018352, 29)

In [143]:
train_m_5 = train_m_4.merge(
    historic_weather_m,
    how="left",
    on=["county", "datetime"],
    suffixes=("", "_hw") 
)
train_m_5.shape

(2018352, 46)

In [144]:
train_m_5.isna().sum()

county                                  0
is_business                             0
product_type                            0
is_consumption                          0
datetime                                0
data_block_id                           0
row_id                                  0
prediction_unit_id                      0
date                                    0
time                                    0
target                                  0
eic_count                               0
installed_capacity                      0
lowest_price_per_mwh                    0
highest_price_per_mwh                   0
euros_per_mwh                           0
temperature                             0
dewpoint                                0
cloudcover_high                         0
cloudcover_low                          0
cloudcover_mid                          0
cloudcover_total                        0
10_metre_u_wind_component               0
10_metre_v_wind_component         

In [145]:
train_m_5[train_m_5.temperature_hw.isna()].date.value_counts()

2023-05-31    3120
2023-05-30    1690
Name: date, dtype: int64

* We do not have historic weather data for **301 and 31** dates. let's imput this with previous day values.

In [146]:
train_m_5[["county", "is_business", "product_type", "datetime", "temperature_hw"]]

county  is_business  product_type            datetime  temperature_hw
0             0            0             1 2021-09-01 00:00:00       13.600000
1             0            0             1 2021-09-01 00:00:00       13.600000
2             0            0             1 2021-09-01 01:00:00       13.566667
3             0            0             1 2021-09-01 01:00:00       13.566667
4             0            0             1 2021-09-01 02:00:00       13.100000
...         ...          ...           ...                 ...             ...
2018347      15            1             3 2023-05-31 21:00:00             NaN
2018348      15            1             3 2023-05-31 22:00:00             NaN
2018349      15            1             3 2023-05-31 22:00:00             NaN
2018350      15            1             3 2023-05-31 23:00:00             NaN
2018351      15            1             3 2023-05-31 23:00:00             NaN

[2018352 rows x 5 columns]

In [147]:
train_m_5[["county", "is_business", "product_type", "is_consumption", "time", "date"]].duplicated().value_counts()

False    2018352
dtype: int64

In [148]:
missing_hw = train_m_5.groupby(["county", "is_business", "product_type", "is_consumption", "time", "date"])["latitude","longitude","temperature_hw","dewpoint_hw","rain","snowfall_hw","surface_pressure","cloudcover_total_hw","cloudcover_low_hw","cloudcover_mid_hw","cloudcover_high_hw","windspeed_10m","winddirection_10m","shortwave_radiation","direct_solar_radiation_hw","diffuse_radiation"].mean().reset_index().ffill()

In [149]:
train_m_5.tail()

county  is_business  product_type  is_consumption  \
2018347      15            1             3               1   
2018348      15            1             3               0   
2018349      15            1             3               1   
2018350      15            1             3               0   
2018351      15            1             3               1   

                   datetime  data_block_id   row_id  prediction_unit_id  \
2018347 2023-05-31 21:00:00            637  2018091                  60   
2018348 2023-05-31 22:00:00            637  2018220                  60   
2018349 2023-05-31 22:00:00            637  2018221                  60   
2018350 2023-05-31 23:00:00            637  2018350                  60   
2018351 2023-05-31 23:00:00            637  2018351                  60   

              date      time  ...  cloudcover_mid_hw  cloudcover_high_hw  \
2018347 2023-05-31  21:00:00  ...                NaN                 NaN   
2018348 2023-05-31  22:00:00  ...                NaN                 NaN   
2018349 2023-05-31  22:00:00  ...                NaN                 NaN   
2018350 2023-05-31  23:00:00  ...                NaN                 NaN   
2018351 2023-05-31  23:00:00  ...                NaN                 NaN   

         windspeed_10m  winddirection_10m  shortwave_radiation  \
2018347            NaN                NaN                  NaN   
2018348            NaN                NaN                  NaN   
2018349            NaN                NaN                  NaN   
2018350            NaN                NaN                  NaN   
2018351            NaN                NaN                  NaN   

         direct_solar_radiation_hw  diffuse_radiation  latitude  longitude  \
2018347                        NaN                NaN       NaN        NaN   
2018348                        NaN                NaN       NaN        NaN   
2018349                        NaN                NaN       NaN        NaN   
2018350                        NaN                NaN       NaN        NaN   
2018351                        NaN                NaN       NaN        NaN   

         data_block_id_hw  
2018347               NaN  
2018348               NaN  
2018349               NaN  
2018350               NaN  
2018351               NaN  

[5 rows x 46 columns]

In [150]:
missing_hw.tail()

county  is_business  product_type  is_consumption      time  \
2018347      15            1             3               1  23:00:00   
2018348      15            1             3               1  23:00:00   
2018349      15            1             3               1  23:00:00   
2018350      15            1             3               1  23:00:00   
2018351      15            1             3               1  23:00:00   

              date  latitude  longitude  temperature_hw  dewpoint_hw  ...  \
2018347 2023-05-27     57.78       27.1            8.62         1.20  ...   
2018348 2023-05-28     57.78       27.1           12.72         4.06  ...   
2018349 2023-05-29     57.78       27.1           11.18         6.98  ...   
2018350 2023-05-30     57.78       27.1           11.18         6.98  ...   
2018351 2023-05-31     57.78       27.1           11.18         6.98  ...   

         surface_pressure  cloudcover_total_hw  cloudcover_low_hw  \
2018347           1010.40                  0.0                0.0   
2018348           1003.16                 17.6                5.2   
2018349           1003.74                 21.2                2.8   
2018350           1003.74                 21.2                2.8   
2018351           1003.74                 21.2                2.8   

         cloudcover_mid_hw  cloudcover_high_hw  windspeed_10m  \
2018347                0.0                 0.0       1.955556   
2018348               21.8                 0.0       3.055556   
2018349               16.4                29.8       1.772222   
2018350               16.4                29.8       1.772222   
2018351               16.4                29.8       1.772222   

         winddirection_10m  shortwave_radiation  direct_solar_radiation_hw  \
2018347              148.6                  0.0                        0.0   
2018348              229.0                  0.0                        0.0   
2018349              170.0                  0.0                        0.0   
2018350              170.0                  0.0                        0.0   
2018351              170.0                  0.0                        0.0   

         diffuse_radiation  
2018347                0.0  
2018348                0.0  
2018349                0.0  
2018350                0.0  
2018351                0.0  

[5 rows x 22 columns]

In [151]:
train_m_5.drop(columns=["latitude", "longitude", "temperature_hw","dewpoint_hw","rain","snowfall_hw","surface_pressure","cloudcover_total_hw","cloudcover_low_hw","cloudcover_mid_hw","cloudcover_high_hw","windspeed_10m","winddirection_10m","shortwave_radiation","direct_solar_radiation_hw","diffuse_radiation","data_block_id_hw"], inplace=True)
train_m_5 = train_m_5.merge(
    missing_hw,
    how="left",
    on=["county", "is_business", "product_type", "is_consumption", "time", "date"],
    suffixes=("", "_hw")
)
train_m_5.shape

(2018352, 45)

In [152]:
train_m_5.isna().sum()

county                               0
is_business                          0
product_type                         0
is_consumption                       0
datetime                             0
data_block_id                        0
row_id                               0
prediction_unit_id                   0
date                                 0
time                                 0
target                               0
eic_count                            0
installed_capacity                   0
lowest_price_per_mwh                 0
highest_price_per_mwh                0
euros_per_mwh                        0
temperature                          0
dewpoint                             0
cloudcover_high                      0
cloudcover_low                       0
cloudcover_mid                       0
cloudcover_total                     0
10_metre_u_wind_component            0
10_metre_v_wind_component            0
data_block_id_fw                     0
direct_solar_radiation   

* No missing data anymore, let's move to next dataset.

In [153]:
del train_m_4

In [155]:
train_m_5.shape

(2018352, 45)

* We have **2018352** records and **45** column in our final dataset. Let's save this for further analysis.

In [154]:
train_m_5.to_pickle(f"{INPUT_DIR}/train_merged.pkl")